src/auth/authenticator/oicd/jwt.go

```diff
@@ -0,0 +1,82 @@
+package oicd
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"crypto/x509"
+	"fmt"
+
+	"github.com/golang-jwt/jwt"
+)
+
+type JWTChecker struct {
+	certs     []*x509.Certificate
+	parser    *jwt.Parser
+	claimsCfg *ClaimsConfig
+}
+
+func NewJWTChecker(certs []*x509.Certificate, claimsCfg *ClaimsConfig, skipClaimsValidation bool) *JWTChecker {
+	return &JWTChecker{
+		certs:     certs,
+		claimsCfg: claimsCfg,
+		parser: &jwt.Parser{
+			SkipClaimsValidation: skipClaimsValidation,
+		},
+	}
+}
+
+func (checker *JWTChecker) Check(_ context.Context, bearerToken string) (*Claims, error) {
+	if len(checker.certs) == 0 {
+		// If no certificate provided we deactivate authentication
+		return nil, nil
+	}
+
+	// Parse and validate token injected in context
+	token, err := checker.parser.ParseWithClaims(
+		bearerToken,
+		&Claims{cfg: checker.claimsCfg},
+		checker.keyFunc,
+	)
+	if err != nil {
+		return nil, err
+	} else if !token.Valid {
+		return nil, fmt.Errorf("invalid access token")
+	}
+
+	return token.Claims.(*Claims), nil
+}
+
+func (checker *JWTChecker) keyFunc(token *jwt.Token) (interface{}, error) {
+	_, ok := token.Method.(*jwt.SigningMethodRSA)
+	if !ok {
+		return nil, fmt.Errorf("invalid access token signing method")
+	}
+
+	for _, cert := range checker.certs {
+		if pubkey, err := tokenAlgoChecker(token.Method.Alg(), cert); err == nil {
+			return pubkey, nil
+		}
+	}
+
+	return nil, fmt.Errorf("unable to find appropriate key in key set.")
+}
+
+func tokenAlgoChecker(algo string, cert *x509.Certificate) (interface{}, error) {
```

Comment:
this function should not exist anymore, you can check the code I sent you. You just need to check the `kid` value. These hardcoded values `"RS256", "RS384", "RS512"` are what `token.Method.(*jwt.SigningMethodRSA)` is doing.

Something like this:
```
_, ok := token.Method.(*jwt.SigningMethodRSA)
		if !ok {
			errMessage := "invalid access token signing method"
			return nil, errors.New(errMessage)
		}
		for _, key := range keySet.Keys {
			if token.Header["kid"] == key.KeyID {
				return key.Certificates[0].PublicKey, nil
			}
		}
```
keySet comes from go-jose

Reply:
I don't want me implementation to be so tight to a third party library. What would it be the advantage on using this go-jose?